### DISCUSSION TITLE
Adapter Maintainers: Upgrading to dbt-core `v1.9.0`

### DISCUSSION THREAD

**Discussion by dbeatty10** [dbt-labs/dbt-core#11125] [Community]
## Overview

This discussion is for communicating to adapter maintainers the scope of work needed to make use of the changes in dbt-core 1.9.0. If you have questions and concerns, please ask them here for posterity.

If there's something missing, please comment below!

<details><summary>

### Release timeline
</summary>

The below table gives the [milestones](https://github.com/dbt-labs/dbt-core/issues/10916) between up to and including the final release.

- [x]  v1.9.0b1 - Oct 1, 2024
- [x]  v1.9.0b2 - Oct 7, 2024
- [x]  v1.9.0b3 - Oct 30, 2024
- [x]  v1.9.0b4 - Nov 6th, 2024
- [x]  v1.9.0rc1 - Nov 25th, 2024
- [x]  v1.9.0rc2 - Dec 2nd, 2024
- [x]  v1.9.0 (final) - Dec 9th, 2024
</details>

<details><summary>

### Prior maintainer upgrade guides
</summary>

- #9798
- #8307
- #7958
- #7213
- #6624
- #6011
- #5468

</details>

### Example Diffs

- https://github.com/dbt-labs/dbt-postgres/compare/v1.8.1...v1.9.0
- https://github.com/dbt-labs/dbt-redshift/compare/v1.8.0...v1.9.0
- https://github.com/dbt-labs/dbt-bigquery/compare/v1.8.0...v1.9.0
- https://github.com/dbt-labs/dbt-snowflake/compare/v1.8.0...v1.9.0
- https://github.com/dbt-labs/dbt-spark/compare/v1.8.0...v1.9.0

## TL;DR

There should be no breaking changes -- we hope and believe that things should work by bumping package and dependency versions -- please let us know if otherwise! We recommend adding tests where possible / appropriate and updating implementations within your adapter as needed to pass those tests -- see below for details.

## Microbatch incremental strategy

See here for a detailed maintainer guide that is specific to microbatches: https://github.com/dbt-labs/dbt-adapters/discussions/371

#### Tests

- [`dbt.tests.adapter.incremental.test_incremental_microbatch.BaseMicrobatch`](https://github.com/dbt-labs/dbt-adapters/blob/fb11e9fa169c21abc1de3888c4c52cb42a308448/dbt-tests-adapter/dbt/tests/adapter/incremental/test_incremental_microbatch.py#L28)

#### Examples how to implement

- Adapters that have already implemented `valid_incremental_strategies` (first available in [1.3](https://github.com/dbt-labs/dbt-core/discussions/6011#discussioncomment-11442118)):
    - [dbt-snowflake#1179](https://github.com/dbt-labs/dbt-snowflake/pull/1179/files)
    - [dbt-redshift#924](https://github.com/dbt-labs/dbt-redshift/pull/924/files)
- Adapters that have **not** already implemented `valid_incremental_strategies`:
    - [dbt-spark#1108](https://github.com/dbt-labs/dbt-spark/pull/1108/files)
    - [dbt-bigquery#1334](https://github.com/dbt-labs/dbt-bigquery/pull/1334/files)

## Snapshot improvements

There are several improvements listed below that give users more control over their snapshots.

If you have any customizations for macros related to dbt snapshots, you will need to update them so they are compatible with the following features.

Additional context is also available in the [epic](https://github.com/dbt-labs/dbt-core/issues/10151).

### Snapshots: Enable setting a datetime value for `dbt_valid_to` for current records instead of `NULL`

For adapters using the default snapshot macros without overrides, this feature will be automatically inherited due to the changes in [dbt-adapters#321](https://github.com/dbt-labs/dbt-adapters/pull/321/files).

The documentation for this feature was added in [docs.getdbt.com#6308](https://github.com/dbt-labs/docs.getdbt.com/pull/6308/files).

#### Test examples

- [`BaseSnapshotDbtValidToCurrent`](https://github.com/dbt-labs/dbt-core/blob/fc6167a2ee5291cf6c562eadcd975b48e6a34d65/tests/functional/adapter/simple_snapshot/test_various_configs.py#L193)

#### Examples how to implement

- [dbt-postgres#152](https://github.com/dbt-labs/dbt-postgres/pull/152/files)
- [dbt-spark#1113](https://github.com/dbt-labs/dbt-spark/pull/1113/files)

### Snapshots: Enable `hard_deletes` config to add a metadata column if a record has been deleted

For adapters using the default snapshot macros without overrides, this feature will be automatically inherited due to the changes in [dbt-adapters#317](https://github.com/dbt-labs/dbt-adapters/pull/317/files).

The documentation for this feature was added in [docs.getdbt.com#6558](https://github.com/dbt-labs/docs.getdbt.com/pull/6558/files).

#### Tests

- `dbt.tests.adapter.simple_snapshot.new_record_mode`
    - https://github.com/dbt-labs/dbt-adapters/blob/fb11e9fa169c21abc1de3888c4c52cb42a308448/dbt-tests-adapter/dbt/tests/adapter/simple_snapshot/new_record_mode.py#L177

### Snapshots: Allow customizing names of metadata fields

For adapters using the default snapshot macros without overrides, this feature will be automatically inherited due to the changes in [dbt-adapters#290](https://github.com/dbt-labs/dbt-adapters/pull/290/files).

The documentation for this feature was added in [docs.getdbt.com#6211](https://github.com/dbt-labs/docs.getdbt.com/pull/6211/files).

#### Test examples

- [`BaseSnapshotColumnNamesFromDbtProject`](https://github.com/dbt-labs/dbt-core/blob/fc6167a2ee5291cf6c562eadcd975b48e6a34d65/tests/functional/adapter/simple_snapshot/test_various_configs.py#L73)
- [`BaseSnapshotColumnNames`](https://github.com/dbt-labs/dbt-core/blob/fc6167a2ee5291cf6c562eadcd975b48e6a34d65/tests/functional/adapter/simple_snapshot/test_various_configs.py#L38)
- [`BaseSnapshotInvalidColumnNames`](https://github.com/dbt-labs/dbt-core/blob/fc6167a2ee5291cf6c562eadcd975b48e6a34d65/tests/functional/adapter/simple_snapshot/test_various_configs.py#L123)

#### Examples how to implement

- [dbt-postgres#145](https://github.com/dbt-labs/dbt-postgres/pull/145)
- [dbt-spark#1097](https://github.com/dbt-labs/dbt-spark/pull/1097)

### Snapshots: Enable `unique_key` as a list

For adapters using the default snapshot macros without overrides, this feature will be automatically inherited due to the changes in [dbt-adapters#326](https://github.com/dbt-labs/dbt-adapters/pull/326/files).

The documentation for this feature was added in [docs.getdbt.com#6438](https://github.com/dbt-labs/docs.getdbt.com/pull/6438/files).

#### Test examples

- [`BaseSnapshotMultiUniqueKey`](https://github.com/dbt-labs/dbt-core/blob/fc6167a2ee5291cf6c562eadcd975b48e6a34d65/tests/functional/adapter/simple_snapshot/test_various_configs.py#L248)

**Comment by damian3031** [comment]
All new tests related to snapshot improvements are added in dbt-core repository. Shouldn't they be added to dbt-adapters repository instead?
Are these new snapshot tests implemented in any adapter maintained by dbt Labs? I couldn’t find any.
